app/controllers/portfolios_controller.rb

```diff
@@ -0,0 +1,99 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+# Controller for items of workspace_type portfolio. Contains special logic that is unique to portfolios.
+#
+# Note that _some_ actions (such as #new) are handled by the project controller since portfolios behave mostly like
+# projects in these cases.
+class PortfoliosController < ApplicationController
+  include OpTurbo::ComponentStream
+
+  include QueriesHelper
+  include Queries::Loading
+
+  authorization_checked! :index
+
+  before_action :authorize_portfolio_access, only: %i[index]
+  before_action :not_authorized_on_feature_flag_inactive
+  before_action :set_default_query, only: %i[index] # Must be called before `load_query_or_deny_access`
+  before_action :load_query_or_deny_access, only: %i[index]
+
+  current_menu_item :index do
+    :portfolios
+  end
+
+  def index # rubocop:disable Metrics/AbcSize
+    respond_to do |format|
+      format.html do
+        flash.now[:error] = @query.errors.full_messages if @query.errors.any?
+
+        render layout: "global", locals: { query: @query, state: :show }
+      end
+
+      format.turbo_stream do
+        replace_via_turbo_stream(
+          component: Portfolios::IndexPageHeaderComponent.new(query: @query, current_user:, params:)
+        )
+        update_via_turbo_stream(
+          component: Projects::ProjectFilterButtonComponent.new(query: @query, disable_buttons: false)
+        )
+        replace_via_turbo_stream(component: Portfolios::IndexComponent.new(query: @query, current_user:))
+
+        current_url = url_for(params.permit(:controller, :action, :query_id, :filters, :sortBy, :page, :per_page))
+        turbo_streams << turbo_stream.push_state(current_url)
+        turbo_streams << turbo_stream.turbo_frame_set_src(
+          "portfolios_sidemenu",
+          portfolios_menu_url(query_id: @query.id, controller_path: "portfolios")
+        )
+
+        turbo_streams << turbo_stream.replace("flash-messages", helpers.render_flash_messages)
+
+        render turbo_stream: turbo_streams
+      end
+    end
+  end
+
+  private
+
+  def set_default_query
+    # When loading the default query (when users enter the page without making a query selection),
+    # we want to ensure that we get a portfolio query. Else, users would see a project query.
+    params[:query_id] ||= ProjectQueries::Static::ACTIVE_PORTFOLIOS
+  end
+
+  def authorize_portfolio_access
+    render_403 unless User.current.allowed_globally?(:add_portfolios) ||
+                       Project.active.portfolio.allowed_to(User.current, :view_project).any?
```

Comment:
The check on `.active` is not necessary here as something is only `allowed_to` in an active project. 

Which brings me to an interesting and as of now undiscussed use case: How do we display archived/inactive portfolios and do we allow activating them?

Currently, we do display inactive portfolios (if filtered for them)

<img width="890" height="466" alt="Image" src="https://github.com/user-attachments/assets/ed170cad-bbd7-43c9-b681-ad2984fdcf8f" />

But clicking on the link will lead to a 403. 

For now I would suggest to have a prefix or some other indication to visually convey that the portfolio is inactive and no longer have a link in this case. I would for now disregard the use case of reactivating the portfolio from this page.